core/services/bulletprooftxmanager/nonce_syncer.go

```diff
@@ -0,0 +1,406 @@
+	// put unnecessary load on the eth node, but the most common use-case is
+	// with a single key and given how rare this scenario is, I think we can
+	// live with it.
+	var reqs2 []rpc.BatchElem
+	for i := floor; i < latestBlockNum; i++ {
+		req := rpc.BatchElem{
+			Method: "getBlockByNumber",
+			Args:   []interface{}{models.Int64ToHex(i), true},
+			Result: &models.Block{},
+		}
+		reqs2 = append(reqs2, req)
+	}
+
+	err = s.ethClient.BatchCallContext(ctx, reqs2)
+	if err != nil {
+		return err
+	}
+
+	reqs = append(reqs, reqs2...)
+	var txes []types.Transaction
+	signer := types.NewEIP155Signer(s.config.ChainID())
+
+	// Rip through all transactions in all blocks and keep only the ones sent
+	// from our key
+	for _, req := range reqs {
+		if req.Error != nil {
+			logger.Warnw("NonceSyncer: got error querying for block", "blockNum", req.Args[0], "err", req.Error)
+			continue
+		}
+		block, is := req.Result.(*models.Block)
+		if !is {
+			panic(fmt.Sprintf("invariant violation, expected %T but got %T", &models.Block{}, block))
+		}
+		for _, tx := range block.Transactions {
+			from, err2 := types.Sender(signer, &tx)
+			if err2 != nil {
+				logger.Warnw("NonceSyncer#fastForwardNonceIfNecessary failed to extract 'from' from transaction", "tx", tx, "err", err2)
+				continue
+			}
+			if from == address {
+				txes = append(txes, tx)
+			}
+		}
+	}
+
+	account, err := s.store.KeyStore.GetAccountByAddress(address)
+	if err != nil {
+		return errors.Wrap(err, "NonceSyncer#fastForwardNonceIfNecessary could not get account from keystore")
+	}
+	sort.Slice(txes, func(i, j int) bool { return txes[i].Nonce() < txes[j].Nonce() })
+
+	inserts, err := s.makeInserts(account, latestBlock.Number, txes, chainNonce-1)
+	if err != nil {
+		return errors.Wrap(err, "NonceSyncer#fastForwardNonceIfNecessary error generating transactions for backfill")
+	}
+
+	now := time.Now()
+	return postgres.GormTransaction(ctx, s.store.DB, func(dbtx *gorm.DB) error {
+		//  We pass in next_nonce here as an optimistic lock to make sure it
+		//  didn't get changed out from under us
+		res := dbtx.Exec(`UPDATE keys SET next_nonce = ?, updated_at = ? WHERE address = ? AND next_nonce = ?`, chainNonce, now, address, localNonce)
+		if res.Error != nil {
+			return res.Error
+		}
+		if res.RowsAffected == 0 {
+			return errors.Errorf("NonceSyncer#fastForwardNonceIfNecessary optimistic lock failure fastforwarding nonce %v to %v for key %s", localNonce, chainNonce, address.Hex())
+		}
+
+		for _, ins := range inserts {
+			// Setting broadcast_at here is a bit of a misnomer since this node
+			// didn't actually broadcast the transaction, but including it
+			// allows us to avoid changing the state machine limitations and
+			// represents roughly the time we read the tx from the blockchain
+			ins.Etx.BroadcastAt = &now
+			if err := dbtx.Create(&ins.Etx).Error; err != nil {
+				return errors.Wrap(err, "NonceSyncer#fastForwardNonceIfNecessary failed to create eth_tx")
```

Comment:
could add 'address' and some other info to the message

Reply:
It shouldn't fail. If so I'd suggest pulling the failure out of the postgres logs would be more useful.

Reply:
ok